prow/pubsub/subscriber/subscriber.go

```diff
@@ -58,6 +59,55 @@ type prowCfgClient interface {
 	GetPostsubmitsStatic(identifier string) []config.Postsubmit
 }
 
+type InRepoConfigCacheGetter struct {
+	CookieFilePath string
+	CacheSize      int
+	Agent          *config.Agent
+
+	CacheMap map[string]*config.InRepoConfigCache
+}
+
+func (irc *InRepoConfigCacheGetter) GetCache(cloneURI, host string) (*config.InRepoConfigCache, error) {
+	if irc.CacheMap == nil {
+		irc.CacheMap = map[string]*config.InRepoConfigCache{}
+	}
+	if cache, ok := irc.CacheMap[cloneURI]; ok {
+		return cache, nil
+	}
+	cache, err := irc.CreateCache(cloneURI, host)
+	if err != nil {
+		return nil, err
+	}
+
+	return cache, nil
+
+}
+
+func (irc *InRepoConfigCacheGetter) CreateCache(cloneURI, host string) (*config.InRepoConfigCache, error) {
+	opts := git.ClientFactoryOpts{
+		CloneURI:       cloneURI,
+		Host:           host,
+		CookieFilePath: irc.CookieFilePath,
+	}
+	gc, err := git.NewClientFactory(opts.Apply)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Gerrit Client for InRepoConfig: %v", err)
+	}
+	// Initialize cache for fetching Presubmit and Postsubmit information. If
+	// the cache cannot be initialized, exit with an error.
+	cache, err := config.NewInRepoConfigCache(
+		irc.CacheSize,
+		irc.Agent,
+		config.NewInRepoConfigGitCache(gc))
+	// If we cannot initialize the cache, exit with an error.
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize in-repo-config-cache with size %d: %v", irc.CacheSize, err)
+	}
+
+	irc.CacheMap[cloneURI] = cache
```

Comment:
sub could process messages in parallel, and this could be executed concurrently, please use mutex to protect map from concurrent racing

Reply:
Done